From 2ee08ec8374d4bfacc14118c34a1bd3392d5fbf2 Mon Sep 17 00:00:00 2001
From: Josselin Mouette <joss@debian.org>
Date: Sun, 23 Sep 2012 11:11:03 +0200
Subject: [PATCH] Display fallback warning using libnotify

This avoids annoying users, especially those using live systems, with a
dialog they have to click on.
---
 configure.ac                |    2 ++
 gnome-session/gsm-manager.c |   61 +++++++++++++++++++------------------------
 2 files changed, 29 insertions(+), 34 deletions(-)

diff --git a/configure.ac b/configure.ac
index b75f269..60a7461 100644
--- a/configure.ac
+++ b/configure.ac
@@ -38,6 +38,7 @@ GTK3_REQUIRED=2.90.7
 DBUS_GLIB_REQUIRED=0.76
 UPOWER_REQUIRED=0.9.0
 JSON_GLIB_REQUIRED=0.10
+LIBNOTIFY_REQUIRED=0.7
 
 dnl ====================================================================
 dnl Dependency Checks
@@ -54,6 +55,7 @@ PKG_CHECK_MODULES(GNOME_SESSION,
         dbus-glib-1 >= $DBUS_GLIB_REQUIRED
         upower-glib >= $UPOWER_REQUIRED
         json-glib-1.0 >= $JSON_GLIB_REQUIRED
+        libnotify >= $LIBNOTIFY_REQUIRED
 )
 
 PKG_CHECK_MODULES(SESSION_PROPERTIES,
diff --git a/gnome-session/gsm-manager.c b/gnome-session/gsm-manager.c
index 9f7cff5..5aebc08 100644
--- a/gnome-session/gsm-manager.c
+++ b/gnome-session/gsm-manager.c
@@ -40,6 +40,8 @@
 
 #include <gtk/gtk.h> /* for logout dialog */
 
+#include <libnotify/notify.h>
+
 #include "gsm-manager.h"
 #include "gsm-manager-glue.h"
 
@@ -1341,46 +1343,37 @@ end_session_or_show_shell_dialog (GsmManager *manager)
 }
 
 static void
+on_link_clicked (NotifyNotification *notif,
+                             char *action,
+                             gpointer data)
+{
+        char *uri = data;
+        gtk_show_uri (NULL, uri, GDK_CURRENT_TIME, NULL);
+}
+
+static gboolean
+notification_show_timeout (gpointer data)
+{
+        NotifyNotification *notif = (NotifyNotification *) data;
+        notify_notification_show (notif, NULL);
+        g_object_unref (G_OBJECT (notif));
+        return FALSE;
+}
+
+static void
 show_fallback_dialog (const char *title,
                       const char *description,
                       const char *link_text,
                       const char *uri)
 {
-        GtkWidget *dialog, *image, *link, *hbox;
-
-        dialog = gtk_message_dialog_new (NULL, 0,
-                                         GTK_MESSAGE_WARNING,
-                                         GTK_BUTTONS_CLOSE,
-                                         "%s", title);
-
-        gtk_window_set_icon_name (GTK_WINDOW (dialog), GSM_ICON_COMPUTER_FAIL);
+        NotifyNotification *notif;
 
-        image = gtk_image_new_from_icon_name (GSM_ICON_COMPUTER_FAIL,
-                                              gsm_util_get_computer_fail_icon_size ());
-        gtk_message_dialog_set_image (GTK_MESSAGE_DIALOG (dialog), image);
-
-        if (description) {
-                gtk_message_dialog_format_secondary_markup (GTK_MESSAGE_DIALOG (dialog),
-                                                            "%s", description);
-        }
-
-        hbox = gtk_message_dialog_get_message_area (GTK_MESSAGE_DIALOG (dialog));
-
-        if (uri) {
-                if (link_text) {
-                        link = gtk_link_button_new_with_label (uri, link_text);
-                } else {
-                        link = gtk_link_button_new (uri);
-                }
-                gtk_box_pack_start (GTK_BOX (hbox), link, FALSE, FALSE, 0);
-        }
-
-        gtk_widget_show_all (dialog);
-
-        g_signal_connect (dialog,
-                          "response",
-                          G_CALLBACK (gtk_widget_destroy),
-                          NULL);
+        notify_init ("GNOME");
+        notif = notify_notification_new (title, description, GSM_ICON_COMPUTER_FAIL);
+        notify_notification_set_timeout (notif, 15*1000);
+        notify_notification_add_action (notif, "link-click", link_text, NOTIFY_ACTION_CALLBACK (on_link_clicked), g_strdup (uri), g_free);
+        /* Give the notification daemon a chance to finish initialization */
+        g_timeout_add_seconds (2, (GSourceFunc) notification_show_timeout, (gpointer) notif);
 }
 
 static void
-- 
1.7.10.4

